File: app/src/main/java/com/github/weatherapp/injection/module/OpenWeatherApiModule.java
package com.github.weatherapp.injection.module;

import com.github.weatherapp.core.WeatherService;
import com.github.weatherapp.data.OpenWeatherApi;
import com.github.weatherapp.data.OpenWeatherService;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;

@Module
public class OpenWeatherApiModule {
    private final String apiKey;

    public OpenWeatherApiModule(String apiKey) {
        this.apiKey = apiKey;
    }

    @Provides
    @Singleton
    static OpenWeatherApi provideOpenWeatherApi(Retrofit retrofit) {
        return retrofit.create(OpenWeatherApi.class);
    }

    @Provides
    @Singleton
    WeatherService provideWeatherService(OpenWeatherApi api) {
        return new OpenWeatherService(api, apiKey);
    }
}

File: app-kotlin/src/main/kotlin/com/github/weatherapp/injection/module/OpenWeatherApiModule.kt
package com.github.weatherapp.injection.module

import com.github.weatherapp.core.WeatherService
import com.github.weatherapp.data.OpenWeatherApi
import com.github.weatherapp.data.OpenWeatherService
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import javax.inject.Singleton

/*
    Open Weather Map API module
 */
@Module
class OpenWeatherApiModule(private val apiKey: String) {

    @Provides
    @Singleton
    fun provideWeatherService(api: OpenWeatherApi): WeatherService {
        return OpenWeatherService(api, apiKey)
    }

    @Provides
    @Singleton
    fun provideOpenWeatherApi(retrofit: Retrofit): OpenWeatherApi {
        return retrofit.create(OpenWeatherApi::class.java)
    }
}
